### ISSUE TITLE
DamageClass overhaul.

### ISSUE THREAD

**Issue by ThomasThePencil** [tModLoader/tModLoader#1173]
The recently-introduced DamageClass feature in tML is a nice way to easily keep track of what classes exist in the game and which items belong to which ones, but it currently lacks some of the deeper functionality some people may expect out of it. My goal is to overhaul this concept into a much more flexible and powerful option for expanding upon the game. The way I can see it currently, there are three potential **major** courses of action regarding the overhaul of DamageClass. I'd like to gather a sufficient number of development-based opinions on these concepts and which one would ultimately be the best for all parties involved (the tML team, the modders, and to a point the players) going into the 1.4 age.

### THE GOAL:
Make DamageClass a little more flexible, and give it further options for the modder to expand upon its concepts, primarily in the field of hybrid items/projectiles and dynamic class bonus scaling.

## ROUTE THE FIRST:
**DamageType is replaced with Dictionaries called DamageTypes** (DamageClass keys, bool values) **and ScalingForDamageTypes** (DamageClass keys, float values)**.** These may be used like so:
```
thing.DamageTypes[DamageClass.Melee] = true;
thing.ScalingForDamageTypes[DamageClass.Melee] = 0.5f;
```
to add a damage class, or multiple of them, to an item.
Class-specific "universal stat modifier" calcs (damage, crit, and (with this overhaul) knockback) will iterate through each entry in DamageTypes and, if that entry's value is true, apply the stat bonuses.
This is the one that I had originally picked for this task, and the one I personally would prefer, as I feel it makes the most sense.

### Pros:
- Allows easy creation of hybrid items and projectiles.
- Presents the setup most similar to the 1.3 method of setting classes for items and projectiles, allowing easier porting.
- Passively corrects the scaling issue with Forbidden armor's Ancient Storm projectile once complete.

### Cons:
- Newer modders might find dictionaries confusing. Moreover, the actual line needed for this is a little long-winded, which may turn off those looking for a simpler way.
- If a DamageClass somehow isn't registered and a mod checks for it, this can result in problems on their end. I feel this is more an issue of modder error, but it's important to bring up.

## ROUTE THE SECOND:
**DamageType is replaced with a List called DamageTypes.** This may be used like so:
```
thing.DamageTypes.Add(DamageClass.Melee);
```
to add a damage class, or multiple of them, to an item.
Class-specific "universal stat modifier" calcs (damage, crit, and (with this overhaul) knockback) will iterate through DamageTypes and apply the stat bonuses for each damage class that is meant to be used for the item.

### Pros:
- Allows easy creation of hybrid items and projectiles.
- Lists are very intuitive to newer modders, and very easy to explain even to those who don't understand them, compared to dictionaries.
- Passively corrects the scaling issue with Forbidden armor's Ancient Storm projectile once complete.

### Cons:
- Arguably the most annoying of the three potential routes to port a 1.3 mod onto.
- Makes dynamic scaling pretty damn ugly to implement on tML's side compared to routes 1 and 3, since not all damage types are guaranteed to be present in the List and thus you have to manually check for each one.
- Not all DamageClasses are registered on projectile creation. As such, the method of finding and removing a given entry (making a projectile NOT a given class) is a little more long-winded compared to dictionaries. Compare:
```
if (thing.DamageTypes.Contains(entry))
    thing.DamageTypes.Remove(entry);
```
to:
```
thing.DamageTypes[(entry)] = false;
```
- If a DamageClass somehow isn't registered and a mod checks for it, this can result in problems on their end. See my previous assertion on this.

## ROUTE THE THIRD:
**DamageType remains a single-DamageClass variable, and the task is simply to add a float called DamageTypeScaling.** These may be used like so:
```
thing.DamageType = DamageClass.Melee;
thing.DamageTypeScaling = 0.5f;
```
to make an item follow one damage class. This is the current behavior of DamageClass with dynamic scaling added, and as such I do not believe it needs further explanation.

### Pros:
- Retains and expands upon the current functionality of DamageClass, thus requiring the least overall work from the current 1.4 branch standpoint to implement.
- Very easy for modders and tML devs alike to keep track of, compared to the other two options.
- Changing a vanilla projectile's damage type is easiest with this setup.

### Cons:
- While not as annoying as route 2 to port to, will still create a large amount of work for those wishing to port their mods from 1.3 to 1.4 (which is a lot of the biggest names in the business, and quite a few smaller names too).
- Requires much more effort, on the modder's side, to create items that scale with multiple damage classes at once. This could be corrected by adding hooks to DamageClass to allow inheriting vanilla damage types, but it's still a lot of needless extra work for the moodder and the modder alone, which counterplays the main intention of giving modders more freedom.
- Requires hardcoding the fix to the issue with Forbidden armor's Ancient Storm projectile (it doesn't actually scale with magic and summon bonuses properly). It'd be rather preferable if we didn't need to hardcode a bugfix, of all things, especially given it's all too possible other modders will add new things that end up with the same problem.

**Comment by Chicken-Bones** [comment]
These suggestions are good, but they're one step too far ahead. You state as context
> but it currently lacks some of the deeper functionality some people may expect out of it. My goal is to overhaul this concept into a much more flexible and powerful option for expanding upon the game

This is what we need to discuss. There was some good discussion in Discord about different weapons and effects people might want to achieve. Your suggestions assume that people want weapons boosted by multiple damage classes, with a scaling modifier on each. 
So you could achieve things like `magicDamage & summonDamage & allDamage`. 
Or `(((allDamage & magicDamage - 1) * 0.5 + 1) * (((allDamage & summonDamage - 1) * 0.5 + 1)`
Honestly given additive and multiplicative damage modifiers, I don't know how your scaling math was going to work.

What I want to see is actual example mod worthy item examples which make sense. Calamity had a bunch. Things with multiple projectiles or swings which do different damage types. Things which benefit from all damageClasses regardless (just & them all together) things which count allDamage multiple times. Etc

**Comment by Mirsario** [comment]
I am against 1st and 2nd options. What is DamageTypeScaling in the 3rd one?

**Comment by ThomasSpeedrunner** [comment]
DamageTypeScaling is meant to be a float (could potentially also be a Modifier since that might be more stable) that adjusts the amount by which the given item's damage type affects the item's damage and crit (a preliminary version of this can already be seen with vanilla's item-specific melee speed scaling system). Weapons which have a built-in progression-based scaling system, such as the Halibut Cannon from Calamity or AlchNPC's *numerous* progression-scaling weapons, could benefit from a feature like this for balancing reasons.

That being said, if options 1 and 2 aren't viable for the overarching goal, what approach would have this sort of elasticity that I'm looking to have added to the concept?
**NOTE:** It'd be best to find an approach that isn't just leaving the modder to fumble around like an idiot for a week wondering why their multiclass item isn't actually multiclass before pulling a hacky-as-fuck solution that ruins mod compat or just scrapping the concept entirely --- because fun fact, you can only use one DamageClass for DamageType currently, and vanilla effects such as Spectre bolts and Shroomite stealth check against that before happening.

**Comment by ThomasSpeedrunner** [comment]
I've been mulling over this for the past few days in my head, and I think I actually have a solution that is not only better than all three of the others, but is also most likely to please the highest number of people. I present to you...

## ROUTE THE FOURTH:
**DamageType remains as is.**

There are two new fields added to DamageClass to facilitate the creation of hybrid items.
**The first is a <DamageClass, float>Dictionary called AffectedBy,** containing by default only the five base classes with values of 0. This dictionary dictates how much a DamageClass is affected by other classes' statistical bonuses (or rather, those covered by DamageClass --- so just damage and crit, and KB once I'm done). The default is that the new DamageClass is not affected by other classes' bonuses at all.
**The second is a <DamageClass>List called CountsAs,** containing nothing by default. This list dictates what classes the new DamageClass counts as for the purpose of effects such as Spectre armor bolts and Frostburn procs from Frost armor.

### Pros:
- Allows easy creation of hybrid items and projectiles without significantly changing the way the system currently works.
- Addresses the issues with all three approaches, the biggest of which is that DamageClass fails to provide the flexibility I was originally looking to have added.
- CountsAs, being a list, is easy to work with even for newer programmers and still addresses the issue of class-specific *effects* not being easy to apply.

### Cons:
- For AffectedBy, as is the case with route 1, newer modders might find dictionaries confusing.

**Comment by Chicken-Bones** [comment]
So, at the risk of sounding like a broken record. Mirsario has provided half of what I asked for.


> * The first one is just multiplying. Calamity has an item that does very low damage, but is multiplied by all class multipliers;
> * The second one is using one maximum multiplier. I can see this being used for throwing weapons, which for some reason are melee in vanilla, i.e. boomerangs;
> * The third one is just having different attacks doing different damage, on LMB and RMB for example. Calamity does have weapons like that, and it may be the most popular thing to do.

I'm given to understand all of these things _have_ been implemented before, (however hacky) and I think they _can_ be implemented now, even more easily. The default calculation just provides a ready to go damage modifier as `allDamage & GetDamage(sItem.DamageType)` and can be completely overwritten to do custom calculations.

What I really want to see is if the current method of implementing any of these features has serious problems that warrant dedicated design support.

For 1. all modifiers, just loop through damage classes, and `&=` the `damage` parameter by all the player's modifiers (you can recount `allDamage` n times if you want, or not completely up to you.

For 2. Similar to 1, just pick the max modifier after casting to float

For 3. How bad is swapping damage type at the moment? This is really only a factor for items which don't spawn projectiles, as you could just change the damage type of the spawned projectile no?

I'd be happy to see an ExampleMod PR demonstrating the 3 options above

**Comment by Ozzatron** [comment]
As the local Calamity representative I feel I should elaborate on all multiclass examples I know of in the mod. They were discussed in broad strokes above and they have been explained in-depth on Discord but it would be good to have them all in one place.

1. [Aether's Whisper](https://calamitymod.gamepedia.com/Aether's_Whisper) -- a switchable-class sniper rifle where left click deals magic damage and right click deals ranged damage. Both the item and projectile need to be able to dynamically change damage types because the same projectile is used for both clicks to prevent redundancy. I expect the magic shot _(and not the ranged shot)_ to benefit from Spectre Armor and similar.
2. [Lunic Eye](https://calamitymod.gamepedia.com/Lunic_Eye) -- an all-class-scaling weapon which simply adds up all 5 class damage modifiers to multiply its damage. This means universal damage boosts such as Avenger Emblem increase its damage by five times as much. This weapon and its projectile _are not any class_, but scale on _every vanilla class at once_. It should not trigger flasks, Spectre Armor, have increased knockback from Hercules Beetle, or any other class-based effect.
3. [Prismatic Breaker](https://calamitymod.gamepedia.com/Prismatic_Breaker) -- a multiclass sword which can fire Charged Blaster Cannon style giant beams, as well as waves of rainbow energy. Both of the projectiles used by this weapon are marked as both `melee` and `ranged`, and the weapon scales based on the average of your melee and ranged damage boosts. That is, it adds half your melee damage and half your ranged damage together and is thus a use case for customizable scaling multipliers. As the projectiles are marked as melee, this weapon benefits only from melee crit (due to the if-else ladder), which displays a weakness of the current system that modders cannot easily rectify. It would be ideal if the weapon averaged together the user's melee and ranged crit chance.
4. [Golden Gun](https://calamitymod.gamepedia.com/Golden_Gun) -- a classless gun whose damage cannot be increased by any means whatsoever. It always does exactly 5 damage and cannot critically strike. Its inability to critically strike is not a specifically intended feature, and thus displays a weakness of the current system. Universal crit bonuses such as Destroyer Emblem are not in fact universal, unlike the `allDamage` added by TML.

**Comment by Chicken-Bones** [comment]
That's a good set of use cases/examples @Ozzatron. From my reading, #1214 makes 2/3/4 easier, and doesn't make 1 harder than it already is.

**Comment by Solxanich** [comment]
Comment: Cleanup of the final damage class design is Release important - summoner bugs remain.
#1741

**Comment by Chicken-Bones** [comment]
Sufficiently implemented in #2196
